### PR TITLE
REPL `str` support

### PR DESCRIPTION
**Pull request by Vipul-Cariappa**

### CODE REVIEW

**Review by Shaikh-Ubaid — COMMENTED** [review]
Two comments. Rest looks good. Thanks for this!

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -1972,6 +1972,10 @@ LFORTRAN_API void _lfortran_strcpy(char** x, char *y, int8_t free_target)
         // *x = (char*) malloc((strlen(y) + 1) * sizeof(char));
         // _lfortran_string_init(strlen(y) + 1, *x);
     }
+    if (y == NULL) {
+        *x = NULL;
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply src/libasr/runtime/lfortran_intrinsics.c:1978]: Without the above change, the following fails (SEGFAULT):

```bash
❯ lp                                                           
>>> s: str
>>> s
The stack address was not found in any shared library or the main program, the stack is probably corrupted. Aborting.
```

Tested by Test Case `PythonCompiler string 2`.

[the remark this answers is not in the document]
Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.c:1978]: I think it is fine. I am curious about the following: 
- why does `_lfortran_strcpy` get called in this case?
- are you able to reproduce this same bug in normal/regular `lpython` invocation?

[the remark this answers is not in the document]
Comment (Vipul-Cariappa) [reply src/libasr/runtime/lfortran_intrinsics.c:1978]: > why does `_lfortran_strcpy` get called in this case?

The `global_stmt` function that is constructed with the top-level expressions, copies the string into the return variable. You can run the example code above with `lp -v` flag to look at the ASR and LLVM IR generated.

The ASR and IR for the second line:
```txt
(TranslationUnit
    (SymbolTable
        1
        {
            __main__:
                (Module
                    (SymbolTable
                        2
                        {
                            __main__global_stmts_2__:
                                (Function
                                    (SymbolTable
                                        3
                                        {
                                            __main__global_stmts_2__1:
                                                (Variable
                                                    3
                                                    __main__global_stmts_2__1
                                                    []
                                                    ReturnVar
                                                    ()
                                                    ()
                                                    Default
                                                    (Character 1 -2 ())
                                                    ()
                                                    BindC
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    __main__global_stmts_2__
                                    (FunctionType
                                        []
                                        (Character 1 -2 ())
                                        BindC
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        .false.
                                    )
                                    []
                                    []
                                    [(Assignment
                                        (Var 3 __main__global_stmts_2__1)
                                        (Var 2 s)
                                        ()
                                    )]
                                    (Var 3 __main__global_stmts_2__1)
                                    Public
                                    .false.
                                    .false.
                                    ()
                                ),
                            s:
                                (Variable
                                    2
                                    s
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Character 1 -2 ())
                                    ()
                                    Interactive
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    __main__
                    []
                    .false.
                    .false.
                )
        })
    []
)
```
```txt
; ModuleID = 'LFortran'
source_filename = "LFortran"

@s = external global i8*

define i8* @__main__global_stmts_2__() {
.entry:
  %__main__global_stmts_2__1 = alloca i8*, align 8
  store i8* null, i8** %__main__global_stmts_2__1, align 8
  %0 = load i8*, i8** @s, align 8
  call void @_lfortran_strcpy(i8** %__main__global_stmts_2__1, i8* %0, i8 0)
  br label %return

return:                                           ; preds = %.entry
  %1 = load i8*, i8** %__main__global_stmts_2__1, align 8
  ret i8* %1
}

declare void @_lfortran_strcpy(i8**, i8*, i8)
```

---

> are you able to reproduce this same bug in normal/regular `lpython` invocation?

No, this does not cause any problem with normal `lpython` execution.

---

Maybe we can optimize it to not make a copy. I will look into it.

**Pull request merged**